File: src/services/discovery/issues.ts
import { listGitHubIssues } from "../workspace/client";
import { aggregateRepositories } from "./aggregate";
import { discoveryRequestLimiter } from "./concurrency";
import { compareDatesDescending, DISCOVERY_SOURCE_PAGE_SIZE, isAssignedOpenIssue } from "./rules";
import type { DiscoveryAggregateResult, DiscoveryAssignedIssue, DiscoveryLoadOptions } from "./types";

export async function loadDiscoveryAssignedIssues(
  repoFullNames: readonly string[],
  options: DiscoveryLoadOptions = {},
): Promise<DiscoveryAggregateResult<DiscoveryAssignedIssue>> {
  const result = await aggregateRepositories(repoFullNames, async (repoFullName) => {
    const issues = await discoveryRequestLimiter.run(() => listGitHubIssues(repoFullName, {
      state: "open",
      assignee: "@me",
      perPage: DISCOVERY_SOURCE_PAGE_SIZE,
      sort: "updated",
      direction: "desc",
    }, { forceRefresh: options.forceRefresh }));
    return {
      items: issues.filter(isAssignedOpenIssue).map((issue) => ({ repoFullName, issue })),
      truncated: issues.length >= DISCOVERY_SOURCE_PAGE_SIZE,
    };
  });
  result.items.sort((left, right) => compareDatesDescending(left.issue.updatedAt, right.issue.updatedAt));
  return result;
}

File: src-tauri/crates/workspace/src/workspace/discovery/scan/issues.rs
use lilia_github_contracts::discovery::{GitHubDiscoveryAssignedIssue, GitHubDiscoverySection};

use super::aggregate::{aggregate_repositories, RepositoryItems, SOURCE_PAGE_SIZE};
use super::time::timestamp;
use crate::runtime::WorkspaceContext as AppHandle;
use crate::workspace::github::github_list_issues;

pub(super) async fn scan_assigned_issues(
    app: AppHandle,
    repositories: Vec<String>,
    force_refresh: Option<bool>,
) -> GitHubDiscoverySection<GitHubDiscoveryAssignedIssue> {
    let mut section = aggregate_repositories(repositories, move |repo_full_name| {
        let app = app.clone();
        async move {
            let issues = github_list_issues(
                app,
                repo_full_name.clone(),
                Some("open".to_string()),
                Some(SOURCE_PAGE_SIZE),
                Some("updated".to_string()),
                Some("desc".to_string()),
                None,
                None,
                Some("@me".to_string()),
                None,
                None,
                None,
                None,
                force_refresh,
            )
            .await?;
            let truncated = issues.len() >= SOURCE_PAGE_SIZE as usize;
            Ok(RepositoryItems {
                items: issues
                    .into_iter()
                    .filter(|issue| issue.state == "open")
                    .map(|issue| GitHubDiscoveryAssignedIssue {
                        repo_full_name: repo_full_name.clone(),
                        issue,
                    })
                    .collect(),
                truncated,
            })
        }
    })
    .await;
    section.items.sort_by(|left, right| {
        timestamp(&right.issue.updated_at).cmp(&timestamp(&left.issue.updated_at))
    });
    section
}
